cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/GoalViolations.java

```diff
@@ -23,73 +23,67 @@
   private final KafkaCruiseControl _kafkaCruiseControl;
-  private final List<Violation> _goalViolations = new ArrayList<>();
+  // The priority order of goals is maintained here.
+  private final List<String> _fixableViolatedGoals;
+  private final List<String> _unfixableViolatedGoals;
   private final boolean _allowCapacityEstimation;
 
   public GoalViolations(KafkaCruiseControl kafkaCruiseControl, boolean allowCapacityEstimation) {
     _kafkaCruiseControl = kafkaCruiseControl;
     _allowCapacityEstimation = allowCapacityEstimation;
+    _fixableViolatedGoals = new ArrayList<>();
+    _unfixableViolatedGoals = new ArrayList<>();
   }
 
   /**
    * Add detected goal violation.
    *
-   * @param priority The priority of the goal.
    * @param goalName The name of the goal.
-   * @param executionProposals Proposals to execute for fixing the detected anomaly. Empty set indicates goal violations
-   *                           due to optimization failures.
+   * @param fixable Whether the violated goal is fixable or not.
    */
-  public void addViolation(int priority, String goalName, Set<ExecutionProposal> executionProposals) {
-    _goalViolations.add(new Violation(priority, goalName, executionProposals));
+  public void addViolation(String goalName, boolean fixable) {
+    if (fixable) {
+      _fixableViolatedGoals.add(goalName);
+    } else {
+      _unfixableViolatedGoals.add(goalName);
+    }
   }
 
   /**
    * Get all the goal violations.
    */
-  public List<Violation> violations() {
-    return _goalViolations;
+  public Set<String> violations() {
+    Set<String> allViolatedGoals = new HashSet<>(_unfixableViolatedGoals);
+    allViolatedGoals.addAll(_fixableViolatedGoals);
+    return allViolatedGoals;
   }
 
   @Override
   public void fix() throws KafkaCruiseControlException {
-    // Fix the violations using a rebalance.
-    try {
-      _kafkaCruiseControl.rebalance(
-          Collections.emptyList(), false, null, new OperationProgress(), _allowCapacityEstimation,
-          null, null, false, null, null);
-    } catch (IllegalStateException e) {
-      LOG.warn("Got exception when trying to fix the cluster. " + e.getMessage());
-    }
-  }
-
-  public static class Violation {
-    private final int _priority;
-    private final String _goalName;
-    private final Set<ExecutionProposal> _executionProposals;
-
-    public Violation(int priority, String goalName, Set<ExecutionProposal> executionProposals) {
-      _priority = priority;
-      _goalName = goalName;
-      _executionProposals = executionProposals;
-    }
-
-    public int priority() {
-      return _priority;
-    }
-
-    public String goalName() {
-      return _goalName;
-    }
-
-    public Set<ExecutionProposal> executionProposals() {
-      return _executionProposals;
+    if (_unfixableViolatedGoals.isEmpty()) {
+      try {
+        // Fix the fixable goal violations with rebalance operation.
```

Comment:
Goals used for (1) detecting the anomalies, and the (2) goals used for fixing them are configured independently. Hence, here we seem to be making the implicit assumption that `_unfixableViolatedGoals` are  subset of goals used for fixing the anomalies. This sounds like a reasonable assumption -- but then we should have a sanity check in (maybe in `KafkaCruiseControlConfig`) to ensure that this is the case by construction.